c/core.go

```diff
@@ -168,18 +185,35 @@ func (cs ChildSpec) Start(
 		// we cancel the childCtx on regular termination
 		defer cancelFn()
 
-		// client logic starts here, and waits until an error (or lack of) is
-		// reported
-		notifyResult(
-			runtimeName,
-			cs.start(childCtx, func(err error) {
-				// we tell the spawner this child thread has started running
-				if err != nil {
-					startCh <- err
-				}
-				close(startCh)
-			}),
-		)
+		// client logic starts here, despite the call here being a "start", we will
+		// block and wait here until an error (or lack of) is reported from the
+		// client code
+		err := cs.start(childCtx, func(err error) {
+			// we tell the spawner this child thread has started running
+			if err != nil {
+				startCh <- err
+			}
+			close(startCh)
+		})
+
+		notification := ChildNotification{
+			runtimeName: runtimeName,
+			err:         err,
+		}
+
+		// We need to notify someone about the notification that got created, it
+		// could either be our parent supervisor (who is reading notifyCh), or, it
+		// could be the client that called the Stop() function, which blocks until a
+		// child finishes. The Stop() _should_ only be called by the supervisor,
+		// this means the supervisor is not going to be on an state where is reading
+		// the notifyCh chan.
```

Comment:
>  Is it valid for a client to call `Stop()`?

The client means in this context, the caller of the function, the only caller should be the Supervisor. I'm planning to move this logic to `internal/c/core.go` so that I can guarantee that and avoid adding it in the comments

> The wording implies handle with care, but I am not sure what I should be careful about

Yep, that's the intent, although I think we should just enforce that at the compiler level if we can (and in this case, golang doesn't get in the way to accomplish it)

> Why does it mean that the supervisor is not going to be on an state where is reading the `notifyCh` chan? Why send it if you know it will not be read?

So, the Supervisor is going to be listening to `notifyCh` after it bootstraps ([here](https://github.com/capatazlib/go-capataz/blob/90dfe7487184829356da5fb918e7d8ef8922a3f1/s/core.go#L281)). It either can be inside this `supervisorLoop` or in the `stopChildren` context. When we cancel the `context.Done` in that loop, that goroutine is no longer reading the `notifyCh`, but rather, is going to be reading the `terminateCh` of each client (via the `client.Stop`) function.

I'm going to improve upon this comment.